docs/proposals/kubelet-qos.md

```diff
@@ -0,0 +1,125 @@
+
+Extension: In the future, we might add a field to PodSpec in the API, to allow users to specify whether a pod is best-effort or top-tier.
+
+## Alternatives Considered
+
+Designing a simple but flexible quality of service framework is very challenging because there are many possible applications that can use tiered pods in different ways. Here are a couple of alternatives considered, and reasons for why we are not implementing them.
+
+One alternative is that we put all best-effort pods into a parent cgroup, and restrict the memory of the parent cgroup based on resource limits specified by the top-tier pods. This alternative would reduce the chance of a system OOM kill because it prevents jobs from using too much memory, but it reduces utilization. Also, the alternative would require require constant adjustment of the memory limit of cgroups (using a user-space control loop), which is tricky, racy, and error-prone. In the proposal outlined in this design document, best-effort pods will be allowed to "eat" into top-tier pods' memory, but will be killed if top-tier pods needs the memory they reserved. This leads to better utilization. It is a difficult tradeoff, but the better utilization of the proposal is why we rejected this alternative.
+
+Another alternative is to implement more tiers, or user-specified numerical priorities that guide Kubelet on which tasks to kill (if jobs are running out of memory, lower priority tasks will be killed). For now, these alternatives will not be implemented because they don’t seem to add much value, could lead to a more cluttered user interface, and are more difficult to implement. Many of these functionalities either involve racy user-space control loops, or kernel modifications.
+
+## Implementation
+
+Implement QoS policy
+- Add a function in kubelet.go that decides whether a pod is best-effort or top-tier
+- For now, return based on whether container limits are specified, but eventually this can be refactored based on a user specified value in the pod spec
+
+Change memory overcommit mode
+- Right now overcommit mode is off on the machines we set up, so if there isn’t enough memory malloc will return null. This prevents QoS, because best-effort pods won’t be evicted. Instead, when there isn’t enough memory, and top-tier pods call malloc, they may not get the memory they want. We want top-tier pods to get the memory they request, and force out best-effort pods.
+- Change the memory overcommit mode to 1, so that using excess memory starts the OOM killer.
+
+Pod admission
+- Modify predicates.CheckPodsExceedingCapacity (in scheduler/algorithm/predicates, which is used by both the Kubelet and the Scheduler)
+- If the pod is best-effort, we admit it automatically
+- Otherwise, admit the top-tier pod if there’s space based on limits of the other top-tier pods
+
+Pod configuration
+- We’re focusing on Docker in this implementation (not RKT)

```

Comment:
From what I understand, the RKT parts of our code are not sophisticated enough to support QoS, and would require changes that are way out of scope. Please feel free to correct me though!